tools/metadatastore-upgrade/main.go

```diff
@@ -0,0 +1,146 @@
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var err error
+		// Retrieve the latest version of Deployment before attempting update
+		// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+		result, err := deploymentsClient.Get(mlMetadataDeployment, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get latest version of Deployment: %v", err)
+		}
+
+		result.Spec.Template.Spec.Containers[0].Image = image
+		result.Spec.Template.Spec.Containers[0].Args = containerArgs
+		_, err = deploymentsClient.Update(result)
+		if err != nil {
+			return err
+		}
+
+		result, err = deploymentsClient.Get(mlMetadataDeployment, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get latest version of Deployment after update: %v", err)
+		}
+
+		waitTime := 0
+		updateSuccessful := false
+		for waitTime < maxWaitTime && !updateSuccessful {
+			if *result.Spec.Replicas != result.Status.ReadyReplicas {
+				time.Sleep(time.Second)
+				waitTime++
+			} else {
+				updateSuccessful = true
+			}
+		}
+
+		if !updateSuccessful {
+			return fmt.Errorf("updated deployment failed to reach running state")
+		}
+
+		return nil
+	})
+}
+
+func main() {
+	var err error
+	flag.Parse()
+	log.SetOutput(os.Stdout)
+
+	if *imageTag == "" {
+		log.Printf("Missing image_tag flag")
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	if *kubeconfigPath == "" {
+		if home := homedir.HomeDir(); home != "" {
+			*kubeconfigPath = filepath.Join(home, ".kube", "config")
+		}
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfigPath)
+	if err != nil {
+		log.Fatalf("Error reading kubeconfig file: %v", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("Error setting up client auth: %v", err)
+	}
+
+	deploymentsClient := clientset.AppsV1().Deployments(*deploymentNamespace)
+
+	originalDeployment, err := deploymentsClient.Get(mlMetadataDeployment, metav1.GetOptions{})
+	if err != nil {
+		log.Fatalf("Failed to get old Deployment: %v", err)
+	}
+
+	originalImage := originalDeployment.Spec.Template.Spec.Containers[0].Image
+	originalContainerArgs := originalDeployment.Spec.Template.Spec.Containers[0].Args
+
+	newImage := mlMetadataImage + ":" + *imageTag
+	log.Printf("Upgrading MetadataStore in Namespace: %s from Image: %s to Image: %s", *deploymentNamespace, originalImage, newImage)
+	err = updateDeployment(deploymentsClient, newImage, append(originalContainerArgs, upgradeFlag))
+	if err == nil {
+		log.Printf("MetadataStore successfully upgraded to Image: %s", newImage)
+		log.Printf("Cleaning up Upgrade")
+		err = updateDeployment(deploymentsClient, newImage, originalContainerArgs)
+		// In a highly unlikely scenario upgrade cleanup can fail.
+		if err != nil {
+			log.Fatalf("Upgrade cleanup failed: %v. \nLikely MetadataStore is in a functioning state.", err)
```

Comment:
At this point, should the user manually attempt to edit the deployment to remove the flag? Also, should we just print this error, and exit successfully, since the upgrade part worked?

Reply:
I am ok taking that approach. Updated code to reflect that.